### PR TITLE
chore: improve CI

### PR DESCRIPTION
**Pull request by TkDodo**
this PR:

- tries to fix most peer dependency warnings
- splits GitHub Actions into multiple steps so that they run in parallel
- moves test execution from one big jest run into the separate packages

### CODE REVIEW

File: packages/solid-query/src/__tests__/createQuery.test.tsx [hunk 1]
```diff
@@ -5191,7 +5149,7 @@ describe('createQuery', () => {
           isFetching,
           isSuccess,
           isStale,
-        })
+        } as any)
```
Comment (TkDodo) [inline packages/solid-query/src/__tests__/createQuery.test.tsx:5152]: @lukesmurray @ardeora any idea why I had to make these changes here? What is the type `Optional` - I got errors that this is not found.

Also, the types we're trying to push into this Array here don't seem to be matching :/

File: packages/solid-query/jest-preset.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+const solidPreset = require('solid-jest/preset/browser/jest-preset')
+const tanStackPreset = require('../../jest-preset')
+
+module.exports = {
+  ...tanStackPreset,
+  ...solidPreset,
+}
```
Comment (TkDodo) [inline packages/solid-query/jest-preset.js:7]: @ardeora @lukesmurray I'm having problems getting the solid tests to work in this PR. They are the only ones that use their own preset. I've merged the solid preset with our own preset, and I think that works.

I've had general troubles with `babel-jest` - it just gave me lots of "unexpected token" errors, as if it didn't transpile jsx / typescript at all. So I switched to ts-jest and things just work. But the solid preset also uses `babel-jest` under the hood as far as I can see, so the errors are still there for the solid package :/

Would appreciate if you could take a look here 🙏

File: .github/workflows/pr.yml [hunk 3]
```diff
@@ -2,33 +2,98 @@ name: pr
...
-        if: matrix.react == '17'
-        env:
-          REACTJS_VERSION: ${{ matrix.react }}
       - name: Upload coverage to Codecov
         uses: codecov/codecov-action@v2
+  lint:
+    name: 'Lint'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (karlhorky) [inline .github/workflows/pr.yml:25]: This can use v3 now, right?

File: packages/solid-query/jest-preset.js (hunk 2 above)
Comment (TkDodo) [reply packages/solid-query/jest-preset.js:7]: I don't really know what I'm doing, but adding a custom transformer (copied from the preset one + added more babel presets) made it work:

21fa9392

File: .github/workflows/pr.yml (hunk 3 above)
Comment (TkDodo) [reply .github/workflows/pr.yml:25]: probably, yes. what's the difference? are there breaking changes?

Comment (karlhorky) [reply .github/workflows/pr.yml:25]: Node 16 by default and a minimum actions runner version: https://github.com/actions/checkout/releases/tag/v3.0.0

File: packages/solid-query/jest-preset.js (hunk 2 above)
Comment (ardeora) [reply packages/solid-query/jest-preset.js:7]: Hello! Apologies for the late reply. I have been travelling internationally this weekend. So I will be AFK till Monday 😅 From a quick look. It seems like it breaks whenever it encounters type definitions/ types in the test file. Maybe @lukesmurray can take a look? If not I'll take a look first thing Monday

File: .github/workflows/pr.yml (hunk 3 above)
Comment (TkDodo) [reply .github/workflows/pr.yml]: done: [42f3c3d](https://github.com/TanStack/query/pull/4288/commits/42f3c3db718340f86f58a45a0f83523b6b7e031a)

File: packages/solid-query/src/__tests__/createQuery.test.tsx (hunk 1 above)
Comment (ardeora) [reply packages/solid-query/src/__tests__/createQuery.test.tsx:5152]: Hello! This is a bug that is caused by an issue upstream. I am writing a small PR to fix all of these any/Partial typings.

Comment (lukesmurray) [reply packages/solid-query/src/__tests__/createQuery.test.tsx:5152]: thank you @ardeora

File: packages/solid-query/jest-preset.js (hunk 2 above)
Comment (lukesmurray) [reply packages/solid-query/jest-preset.js:7]: let me know if you still need help with this. I'm not super familiar with solid's babel + jest setup but I can do my best to help if you're still running into issues.

Comment (TkDodo) [reply packages/solid-query/jest-preset.js:7]: it works fine now, so this is resolved for me 🤷

File: packages/solid-query/src/__tests__/createQuery.test.tsx (hunk 1 above)
Comment (ardeora) [reply packages/solid-query/src/__tests__/createQuery.test.tsx:5152]: No worries! #4308 aims to solve this and removes all the vague typings too

File: .github/workflows/pr.yml [hunk 4]
```diff
@@ -2,33 +2,98 @@ name: pr
...
+    name: 'Test React 17'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: pnpm/action-setup@v2.2.2
+        with:
+          version: 7
+      - uses: actions/setup-node@v3
+        with:
+          node-version: ${{ matrix.node }}
```
Comment (DamianOsipiuk) [inline .github/workflows/pr.yml:76]: I guess this matrix could be swapped to specific version, like in other jobs?

**Review by DamianOsipiuk — APPROVED** [review]

File: .github/workflows/pr.yml (hunk 4 above)
Comment (TkDodo) [reply .github/workflows/pr.yml]: well spotted - this was a left-over: dc8819cc

**Pull request merged**